### ISSUE TITLE
[RFC] Add Intel Quantization Support in SGLang

### ISSUE THREAD

**Issue by mengniwang95** [sgl-project/sglang#15725]
## Motivation
Previously, we contributed to SGLang by enabling the loading and inference of AutoRound-quantized models. Building on this groundwork, we propose to add an unified quantization support by integrating the AutoRound quantization algorithm into SGLang.

As an advanced algorithm in Intel® Neural Compressor, AutoRound delivers exceptional results across 2 to 4 bits. With AutoRound’s quantization capabilities integrated into SGLang, we can:
- Enhance both online and offline quantization for SGLang
- Provide users seamless access to state-of-the-art quantization that balances speed, accuracy, and ease of use
- Enable future formats (MXFP4/MXFP8/NVFP4, and advanced mixed-precision recipes)

### One-step quantization (online)
Quantize the model during model loading. 

#### Crucial code change:

1. Define a new ModelLoader to get quantized weight online:
```python
class INCModelLoader(DefaultModelLoader):

    @staticmethod
    def load_weights_and_postprocess(model, weights, target_device):
        from auto_round import AutoRound

        autoround = AutoRound(model.config.name_or_path, scheme=model.quant_config.scheme, iters=0)
        autoround.formats = model.quant_config.formats
        autoround.inplace = True
        autoround.immediate_packing = True
        autoround.quantize()
        model.load_weights(iter(autoround.model.state_dict().items()))

        for _, module in model.named_modules():
            quant_method = getattr(module, "quant_method", None)
            if quant_method is not None:
                with device_loading_context(module, target_device):
                    quant_method.process_weights_after_loading(module)
                if _is_npu:
                    torch.npu.empty_cache()
```

2. Add config for quant_method mapping:
```python
class INCWNA16Config(QuantizationConfig):
    def __init__(
        self,
        exclude_modules=None,
    ) -> None:
        ...

    def get_quant_method(
        self,
        layer: torch.nn.Module,
        prefix: str,
    ):
        if isinstance(layer, LinearBase):
            return AWQLinearMethod(self)
        elif isinstance(layer, FusedMoE):
            return MoeWNA16Method(self)
        return None
```
For SGLang existing quant method, we can reuse then.

#### Usage
```bash
python3 -m sglang.launch_server --model-path TinyLlama/TinyLlama-1.1B-Chat-v1.0 --port 30000 --host 0.0.0.0 --quantization autoround_w4a16
```

### Two-step quantization (offline)
Use SGLang's model_loader API to quantize and save the model, then use SGLang to load the quantized model.

#### Crucial code change:
Define a new ModelLoader to quantize and save model:

```python
class INCModelLoader(DefaultModelLoader):

    def __init__(self, load_config: LoadConfig):
        super().__init__(load_config)

    def load_model(
        self,
        *,
        model_config: ModelConfig,
        device_config: DeviceConfig,
    ) -> nn.Module:

       # invoke autoround to do quantization/tuning and saving
       from auto_round import AutoRound
       autoround = AutoRound(model_config.model_path, **quant_kwargs)
       autoround.quantize_and_save(
            output_dir=self.load_config.inc_save_path,
            format=self.load_config.inc_save_formats,
       )
       
       return autoround.model
```
#### Steps to get and run the quantized model:

1. Quantize and save the model
```python
from sglang.srt.configs.load_config import LoadConfig
from sglang.srt.configs.model_config import ModelConfig
from sglang.srt.model_loader.loader import get_model_loader

# Configure model with inc quantization and saving
model_config = ModelConfig(
    model_path="TinyLlama/TinyLlama-1.1B-Chat-v1.0",
    quantization="autoround_w4a16",
    trust_remote_code=True,
)

load_config = LoadConfig(
    inc_save_path="./quantized_model",
)

# Load and quantize the model
model_loader = get_model_loader(load_config, model_config)
quantized_model = model_loader.load_model(
    model_config=model_config,
)
```

2. Load the quantized model
```bash
python3 -m sglang.launch_server --model-path quantized_model/ --quantization auto-round --port 30000 --host 0.0.0.0
```

### Summary
- For online quantization, one-step with auto-round RTN quantization is recommended.
- For offline quantization, two-step quantization are recommended.

#### Comparison between Online & Offline quantization
|  | Online quantization | Offline quantization |	
|---------|---------|---------|
| Support usage | RTN/Optimized_RTN | Tuning/RTN |
| VRAM Usage | Low | High (Tuning) |
| Accuracy | Basic | Better |
| Time Cost | Low | High (Tuning) |

### Timeline
The PR to SGLang is planned to be ready by mid-January 2026.

**Comment by hshen14** [comment]
CC @mingfeima

**Comment by hshen14** [comment]
@AniZpZ Please help review. Thanks.
